### DISCUSSION TITLE
Why was std_list.i removed from the PHP compiler?

### DISCUSSION THREAD

**Discussion by pstephan1187** [swig/swig#2529] [Q&A] [answered]
Hello! I am attempting to use swig to create PHP bindings for a C++ library. The library has the swig `.i` files all ready to go. In one of their `.i` files, they import the `std_list.i` file. That file (among many others) does not exist in the `Lib/php/` directory, but does in others, like python, it does. When I attempt to use swig to compile my target library down to a php extension, I get the following error:

```
templates.i:3: Error: Unable to find 'std_list.i'
make: *** [Makefile:965: phpswig] Error 1
```

Here is the command that is run when that error is thrown:
```
/usr/bin/swig -php -c++ -o php/Sword.cxx -I. -I/usr/include/sword ./sword.i
```

The `templates.i` file includes the `std_list.i` file, throwing that error on compilation.

I am using the swig binary that is installed via `apt-get` in ubuntu 20.04. `swig -version` prints the following:
```
SWIG Version 4.0.1

Compiled with g++ [aarch64-unknown-linux-gnu]

Configured options: +pcre

Please see http://www.swig.org for reporting bugs and further information
```

I am attempting to use this with php7.4

Here is the library I am trying to compile: https://wiki.crosswire.org/DevTools:SWORD
It's php bindings are very old and designed for php4. I have gone through their configuration and adjusted it as best I can for php7.4. Swig does attempt to compile, but like I said, runs into issue with that missing list file. I am unsure if there are other files missing at this point.

So I guess I have a couple questions:
1. Why is the `std_list.i` file not part of the php "driver"?
2. How would I resolve this issue?

Thank you very much for any help.

**Comment by ojwb** [comment, marked as answer]
I don't believe SWIG/PHP ever supported `std_list.i`.  I can find no trace in git anyway

```
$ git log -- Lib/{php,php4,php5,php7}/std_list.i
$
```

(The library directory name has evolved as PHP versions have come and gone.)

> How would I resolve this issue?

If you want to help resolve, you could come up with a patch to implement it - it'll presumably look fairly similar to `Lib/php/std_vector.i` but without things like subscripting.  I'd suggest looking at `std_list.i` vs `std_vector.i` for other languages to see how they compare.

There should really be test coverage for it too.  There's an existing `li_std_list` testcase, so it needs a suitable `Examples/test-suite/php/li_std_list_runme.php` writing (look at other `_runme.php` and also `Examples/test-suite/*/li_std_list_runme.*` to see what other target languages test for this testcase).
